Feature/database ci

adrianprelipcean opened the pull request:
Added continuous integration pipeline for the database:
- linting via `sqlfluff`
- testing via `pgTAP` 

Added unit and performance testing.

Swapped raw SQL access to stored procedures access from within stores.

adrianprelipcean committed:
meta: Added sql linting to CI.

adrianprelipcean committed:
fix: decrement calls correct query

adrianprelipcean committed:
meta: updated linting to include database folder

adrianprelipcean committed:
test: added database test specs

adrianprelipcean committed:
test: fixed node test on sql query assertion

adrianprelipcean committed:
Feature/enable db tests (#6)

meta: enabled database tests

adrianprelipcean committed:
feat: Changed raw SQL calls to stored procedures.

adrianprelipcean committed:
test: changed database performance teststo reference functions

adrianprelipcean committed:
test: added database unit tests for aggregated store

adrianprelipcean committed:
test: added database unit tests for individual store

adrianprelipcean committed:
test: added database unit tests for session handling

adrianprelipcean committed:
Merge pull request #7 from express-rate-limit/feature/db-stored-procedures

Feature/db stored procedures

adrianprelipcean committed:
docs: updated tree structure for contributing.md

adrianprelipcean committed:
docs: updated changelog

adrianprelipcean committed:
1.3.0

nfriedly commented on .github/workflows/cicd.yml:
```diff
@@ -20,10 +20,93 @@ jobs:
+        cpm install TAP::Parser::SourceHandler::pgTAP
+      env:
+        SHELL: /bin/bash
+    - name: Initialize database and run tests
+      run: |
+        bash db/cli/init_db.sh
+        bash db/cli/apply_all_migrations.sh
+        bash db/cli/run_tests.sh
```
I might break these into 3 steps just to make it easier to narrow down if one failed.

nfriedly commented on db/cli/apply_all_migrations.sh:
```diff
@@ -0,0 +1,8 @@
+echo 'Applying migrations'
+for f in source/migrations/*.sql;
```
This makes me a little nervous - the sort order is fairly consistent, but there are things that can affect it. Also, it's pretty dumb - if the project ever gets 10+ migrations, I believe this will run # 1, then # 10, then # 2... (Or maybe it starts with # 10 and then # 1?)

I believe this would work:

```suggestion
for f in `ls -v source/migrations/*.sql`;
```

Alternatively, you could make a small command-line wrapper around [migration_handler.ts](https://github.com/express-rate-limit/rate-limit-postgresql/blob/61ce69d6f6b2921db674f744cf0ae4bbbd68a6a9/source/util/migration_handler.ts) and execute that instead, or even just count on the library itself running them during the first test. (Although, if this is just supposed to be a test of the migration scripts, then I kind of like the bash version better for it's simplicity. You could always drop the tables and do a second test that runs the migrations via node to cover that case.)

adrianprelipcean replied:
Thanks, I completely missed this, I switched to leaving the first init script with `1-` as prefix, and moved the rest to the standard of using `unix-time-` as a prefix. I also added a script for generating a new migration for the new prefix logic.

adrianprelipcean replied:
Had to switch back to incremental prefixes due to requirements from [postgres-migrations](https://github.com/thomwright/postgres-migrations#file-name)

nfriedly commented on db/linting/lint.sh:
```diff
@@ -0,0 +1,5 @@
+echo 'Linting migrations folder'
+sqlfluff lint source/migrations/ --dialect postgres --ignore parsing --config db/linting/.sqlfluff 
+
+echo 'Linting tests folder'
+sqlfluff lint db/tests/**/ --dialect postgres --ignore parsing --config db/linting/.sqlfluff 
```
Nice. *Of course* you can lint sql files. Never even occurred to me...

nfriedly reviewed (approved):
This looks pretty good to me, although I've never done anything too serious with postgresql, so pgTAP and the like are brand new to me. Still the parts I looked up all seemed correct.

adrianprelipcean committed:
meta: improved db migrations for ci.

Added create_migration script.
Forced natural ordering in apply_all_migrations script.

adrianprelipcean committed:
meta: split steps for ci-db-testing

adrianprelipcean committed:
meta: reverted ci-db-testing

adrianprelipcean committed:
hotfix: allow for same key to be present in multiple aggregated sessions

adrianprelipcean committed:
test: added tests for same key being present in multiple sessions

adrianprelipcean committed:
fix: switched back to incremental database migrations
